File: bindings/python/sop/examples/relations_demo.py
import sop.context as context
import sop.database as database
import sop.transaction as transaction
from sop.btree import BtreeOptions, Relation
import os
from dataclasses import dataclass, asdict

@dataclass
class User:
    id: str
    name: str

@dataclass
class Order:
    id: str
    user_id: str
    amount: float

def run():
    # 1. Setup
    ctx = context.Context()
    # Define database options pointing to "data" folder (must be absolute or relative path)
    opts = database.DatabaseOptions(stores_folders=["data"])
    db = database.Database(opts)
    
    # 2. Begin Transaction
    t = db.begin_transaction(ctx, mode=transaction.TransactionMode.ForWriting.value)
    
    try:
        # 3. Create 'Users' Store (Target)
        print("Creating 'users' store...")
        user_store = db.new_btree(ctx, "users", t)
        user_store.add(ctx, "user_1", asdict(User(id="user_1", name="Alice")))
        
        # 4. Create 'Orders' Store with Relation Metadata (Source)
        print("Creating 'orders' store with Relation metadata...")
        
        # Define the relationship: orders.user_id -> users.id
        rel = Relation(
            source_fields=["user_id"],
            target_store="users",
            target_fields=["id"]
        )
        
        opts = BtreeOptions(
            relations=[rel]
        )
        
        # Note: We expect the Value type to be a dict/object containing 'user_id'
        order_store = db.new_btree(ctx, "orders", t, options=opts)
        
        # Add a dummy order using strongly-typed dataclass
        order = Order(id="order_A", user_id="user_1", amount=100)
        order_store.add(ctx, "order_A", asdict(order))

        # 5. Commit
        t.commit(ctx)
        print("Successfully created stores with Relations!")
        print("Verified: Order Store 'orders' is linked to 'users'.")

    except Exception as e:
        t.rollback(ctx)
        print(f"Error: {e}")
        raise

if __name__ == "__main__":
    run()

File: bindings/csharp/Sop.CLI/RelationsDemo.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Sop.CLI;

public class User
{
    [JsonPropertyName("id")]
    public string Id { get; set; }
    [JsonPropertyName("name")]
    public string Name { get; set; }
}

public class Order
{
    [JsonPropertyName("id")]
    public string Id { get; set; }
    [JsonPropertyName("user_id")]
    public string UserId { get; set; }
    [JsonPropertyName("amount")]
    public decimal Amount { get; set; }
}

public class RelationsDemo
{
    public static void Run()
    {
        Console.WriteLine("Running Relations Demo...");
        // 1. Setup
        var ctx = new Context();
        var opts = new DatabaseOptions
        {
            StoresFolders = new List<string> { "data" }
        };
        
        // Ensure directory exists
        System.IO.Directory.CreateDirectory("data");
        
        var db = new Database(opts);

        using (var trans = db.BeginTransaction(ctx))
        {
            // 2. Create 'Users' Store (Target)
            Console.WriteLine("Creating 'users' store...");
            // Use strongly-typed User class
            var users = db.NewBtree<string, User>(ctx, "users", trans);
            users.Add(ctx, "user_1", new User { Id = "user_1", Name = "Alice" });

            // 3. Create 'Orders' Store with Relation Metadata (Source)
            Console.WriteLine("Creating 'orders' store with Relation metadata...");
            
            var rel = new Relation(
                new List<string> { "user_id" }, 
                "users", 
                new List<string> { "id" }
            );

            var options = new BtreeOptions("orders")
            {
                Relations = new List<Relation> { rel }
            };

            // Use strongly-typed Order class
            var orders = db.NewBtree<string, Order>(ctx, "orders", trans, options);
            
            // Add a dummy order using strongly-typed object
            orders.Add(ctx, "order_A", new Order { Id = "order_A", UserId = "user_1", Amount = 100 });

            trans.Commit();
            Console.WriteLine("Successfully created stores with Relations!");
        }
    }
}
